crates/burn-optim/src/optim/muon.rs

```diff
@@ -0,0 +1,787 @@
+    }
+}
+
+/// Parameters for Newton-Schulz orthogonalization.
+#[derive(Clone, Copy)]
+struct NewtonSchulzParams {
+    a: f32,
+    b: f32,
+    c: f32,
+    steps: usize,
+}
+
+impl NewtonSchulzParams {
+    fn new(coefficients: (f32, f32, f32), steps: usize) -> Self {
+        Self {
+            a: coefficients.0,
+            b: coefficients.1,
+            c: coefficients.2,
+            steps,
+        }
+    }
+}
+
+/// Muon optimizer.
+///
+/// Muon internally runs standard SGD-momentum, and then performs an orthogonalization
+/// post-processing step, in which each 2D parameter's update is replaced with the
+/// nearest orthogonal matrix. For efficient orthogonalization we use a Newton-Schulz
+/// iteration, which has the advantage that it can be stably run in bfloat16 on the GPU.
+///
+/// # Important Notes
+///
+/// 1. **Only for 2D+ parameters**: Muon is designed for weight matrices. Use AdamW
+///    or SGD for biases, embeddings, and layer norms.
+///
+/// 2. **Learning rate adjustment**: Muon automatically adjusts the learning rate based
+///    on parameter shape. See [`AdjustLrFn`] for details.
+///
+/// 3. **Weight decay timing**: Unlike typical optimizers, Muon applies weight decay
+///    AFTER orthogonalization but uses the original (unadjusted) learning rate for it.
+#[derive(Clone)]
+pub struct Muon<B: Backend> {
+    momentum: Momentum<B>,
+    ns_params: NewtonSchulzParams,
+    weight_decay_penalty: Option<f32>,
+    epsilon: f32,
+    adjust_lr_fn: AdjustLrFn,
+}
+
+impl<B: Backend> Muon<B> {
+    /// Adjust learning rate based on parameter shape.
+    ///
+    /// # Arguments
+    ///
+    /// * `lr` - Base learning rate
+    /// * `shape` - Parameter shape (uses first two dimensions)
+    ///
+    /// # Returns
+    ///
+    /// Adjusted learning rate
+    ///
+    /// ```ignore
+    /// // For a [1024, 512] weight matrix with lr=0.01:
+    /// // Original: 0.01 * sqrt(1024/512) = 0.01 * 1.414 = 0.01414
+    /// // MatchRmsAdamW: 0.01 * 0.2 * sqrt(1024) = 0.01 * 0.2 * 32 = 0.064
+    /// ```
+    fn adjust_lr(&self, lr: LearningRate, shape: &[usize]) -> LearningRate {
+        lr * self.adjust_lr_fn.adjustment_ratio(shape)
+    }
+
+    /// Perform Newton-Schulz orthogonalization on a gradient tensor.
+    ///
+    /// This computes the zeroth power (orthogonalization) of the input matrix G
+    /// using a quintic Newton-Schulz iteration.
+    ///
+    /// # Algorithm
+    ///
+    /// 1. Transpose if tall matrix (A > B)
+    /// 2. Normalize: X = X / ||X||
+    /// 3. For k steps:
+    ///    - A = X @ X^T
+    ///    - B = b*A + c*A^2
+    ///    - X = a*X + B@X
+    /// 4. Transpose back if needed
+    ///
+    /// # References
+    ///
+    /// - Original: https://github.com/KellerJordan/Muon/blob/master/muon.py
+    /// - PyTorch: https://github.com/pytorch/pytorch/blob/main/torch/optim/muon.py
+    fn zeropower_via_newtonschulz<const D: usize>(&self, g: Tensor<B, D>) -> Tensor<B, D> {
+        assert!(
```

Comment:
Unsure if this should be the default behavior. There isn't a great way yet to define multiple optimizers for a single burn module (e.g a linear layer with a bias vector). Do you have an idea: @laggui ?